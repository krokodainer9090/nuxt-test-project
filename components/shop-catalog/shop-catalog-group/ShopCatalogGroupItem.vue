<script>
export default {
  name: 'ShopCatalogGroupItem',

  props: {
    item: {
      type: Object,
      default: _ => ({})
    }
  },

  data: () => ({
    isPriceIncreased: false,
    isSamePrice: false,
    canSetClasses: false
  }),

  computed: {
    itemPriceClasses () {
      const { isPriceIncreased, isSamePrice, canSetClasses } = this;

      return {
        'shop-catalog-group-item__price-container_increased': canSetClasses && isPriceIncreased && !isSamePrice,
        'shop-catalog-group-item__price-container_decreased': canSetClasses && !isPriceIncreased && !isSamePrice
      };
    },

    itemFullName () {
      const { N, count } = this.item;

      return N && count
        ? `${N}(${count})`
        : '';
    },

    itemPrice () {
      return this.item?.price || '';
    }
  },

  watch: {
    itemPrice (newPrice, oldPrice) {
      this.canSetClasses = true;

      this.isSamePrice = newPrice === oldPrice;
      this.isPriceIncreased = newPrice > oldPrice;
    }
  },

  methods: {
    clickItem () {
      this.$emit('on-click-item', this.item);
    }
  }
};
</script>

<template>
  <div class="shop-catalog-group-item">
    <div
      class="shop-catalog-group-item__content"
      @click="clickItem"
    >
      <span
        v-if="itemFullName"
        class="shop-catalog-group-item__name"
      >
        {{ itemFullName }}
      </span>
      <div
        :class="itemPriceClasses"
        class="shop-catalog-group-item__price-container"
      >
        <span
          v-if="itemPrice"
          class="shop-catalog-group-item__price"
        >
          {{ item.price }}
        </span>
      </div>
    </div>
  </div>
</template>

<style lang="scss">
@import "../../../styles/components/shop-catalog/shop-catalog-group/shop-catalog-group-item";
</style>
