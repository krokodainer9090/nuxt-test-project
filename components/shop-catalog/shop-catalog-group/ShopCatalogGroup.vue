<script>
// Components
import ShopCatalogGroupItem from './ShopCatalogGroupItem';

export default {
  name: 'ShopCatalogGroup',

  components: {
    ShopCatalogGroupItem
  },

  props: {
    group: {
      type: Object,
      default: _ => ({})
    }
  },

  computed: {
    groupItems () {
      return this.group?.items || [];
    }
  },

  methods: {
    getGroupItemKey (key) {
      return `shop-catalog-group-item-${key}`;
    },

    onClickItem (item) {
      this.$emit('on-click-item', item);
    }
  }
};
</script>

<template>
  <div class="shop-catalog-group">
    <h3 class="shop-catalog-group__title">{{ group.G }}</h3>
    <div class="shop-catalog-group__content container">
      <div class="row">
        <ShopCatalogGroupItem
          v-for="(item, index) in groupItems"
          :key="getGroupItemKey(item.id || index)"
          :item="item"
          class="shop-catalog-group__group-item col-12 col-xl-6"
          @on-click-item="onClickItem"
        />
      </div>
    </div>
  </div>
</template>

<style lang="scss">
@import "../../../styles/components/shop-catalog/shop-catalog-group/shop-catalog-group.scss";
</style>
