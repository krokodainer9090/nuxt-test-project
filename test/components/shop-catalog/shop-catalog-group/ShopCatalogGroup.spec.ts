// Components
import ShopCatalogGroup from '~/components/shop-catalog/shop-catalog-group/ShopCatalogGroup.vue';
import ShopCatalogGroupItem from '~/components/shop-catalog/shop-catalog-group/ShopCatalogGroupItem.vue';

// Utilities
import { shallowMount, Wrapper } from '@vue/test-utils';
import { shopCatalogGroup, shopCatalogItem } from '~/test/utils/misc/constants';

describe('ShopCatalogGroup', () => {
  const mountFunction: (options?: Object) => Wrapper<any> = (options = {}) => {
    return shallowMount(ShopCatalogGroup, {
      ...options
    });
  };

  it('should render empty group', () => {
    const wrapper = mountFunction();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render group with items', () => {
    const wrapper = mountFunction({
      propsData: {
        group: shopCatalogGroup
      }
    });
    const emitObject = wrapper.emitted();

    expect(wrapper.html()).toMatchSnapshot();

    // Test emit on-click-item
    wrapper.findComponent(ShopCatalogGroupItem).vm.$emit('on-click-item', shopCatalogItem);

    expect(emitObject['on-click-item']?.[0]).toEqual([shopCatalogItem]);
  });
});
