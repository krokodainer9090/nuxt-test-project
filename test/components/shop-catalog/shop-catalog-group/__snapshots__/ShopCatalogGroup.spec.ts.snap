// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShopCatalogGroup should render empty group 1`] = `
"<div class=\\"shop-catalog-group\\">
  <h3 class=\\"shop-catalog-group__title\\"></h3>
  <div class=\\"shop-catalog-group__content container\\">
    <div class=\\"row\\"></div>
  </div>
</div>"
`;

exports[`ShopCatalogGroup should render group with items 1`] = `
"<div class=\\"shop-catalog-group\\">
  <h3 class=\\"shop-catalog-group__title\\">Группа 1</h3>
  <div class=\\"shop-catalog-group__content container\\">
    <div class=\\"row\\">
      <shopcataloggroupitem-stub item=\\"[object Object]\\" class=\\"shop-catalog-group__group-item col-12 col-xl-6\\"></shopcataloggroupitem-stub>
      <shopcataloggroupitem-stub item=\\"[object Object]\\" class=\\"shop-catalog-group__group-item col-12 col-xl-6\\"></shopcataloggroupitem-stub>
      <shopcataloggroupitem-stub item=\\"[object Object]\\" class=\\"shop-catalog-group__group-item col-12 col-xl-6\\"></shopcataloggroupitem-stub>
    </div>
  </div>
</div>"
`;
